File: 3449.maximize-the-minimum-game-score.java
#
# @lc app=leetcode id=3449 lang=java
#
# [3449] Maximize the Minimum Game Score
#

# @lc code=start
class Solution {
    public long maxScore(int[] points, int m) {
        int n = points.length;
        long left = 0, right = (long)1e15, answer = 0;
        while (left <= right) {
            long mid = left + (right - left) / 2;
            if (canAchieve(points, m, mid)) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return answer;
    }

    private boolean canAchieve(int[] points, int m, long target) {
        long moves = 0;
        for (int p : points) {
            moves += (target + p - 1) / p; // ceil(target/p)
            if (moves > m) return false;
        }
        return moves <= m;
    }
}
# @lc code=end

File: 3449.maximize-the-minimum-game-score.go
#
# @lc app=leetcode id=3449 lang=golang
#
# [3449] Maximize the Minimum Game Score
#
# @lc code=start
func maxScore(points []int, m int) int64 {
    n := len(points)
    left, right := 0, int(1e18) // Setting high right bound due to constraints
    
    check := func(minVal int64) bool {
        moves := int64(0)
        for _, point := range points {
            if point < minVal {
                moves += minVal - int64(point)
            }
        }
        return moves <= int64(m)
    }
    
    for left < right { // Binary search for maximum possible minimum score
        mid := (left + right + 1) / 2
        if check(mid) {
            left = mid // Try for a higher minimum score if possible
        } else {
            right = mid - 1 // Reduce the maximum possible score when condition fails
        }
    }
    return int64(left) // This will be our answer since it maximizes our minimum game score.
}
# @lc code=end